File: Advanced_Robotics_Course/Chapter20/Lesson5/Chapter20_Lesson5.cpp
#include <iostream>
#include <vector>
#include <cmath>

struct EpisodeMetrics {
    int success;
    double time;
    double energy;
};

struct EvalStats {
    double p_hat;
    double p_ci_low;
    double p_ci_high;
    double time_mean;
    double time_std;
    double energy_mean;
    double energy_std;
};

EvalStats compute_stats(const std::vector<EpisodeMetrics>& data) {
    const std::size_t N = data.size();
    double sum_success = 0.0;
    double sum_time = 0.0, sum_time2 = 0.0;
    double sum_energy = 0.0, sum_energy2 = 0.0;

    for (const auto& e : data) {
        sum_success += static_cast<double>(e.success);
        sum_time += e.time;
        sum_time2 += e.time * e.time;
        sum_energy += e.energy;
        sum_energy2 += e.energy * e.energy;
    }

    EvalStats out{};
    out.p_hat = sum_success / static_cast<double>(N);

    const double z = 1.96; // 95% CI
    const double se_p = std::sqrt(out.p_hat * (1.0 - out.p_hat) / static_cast<double>(N));
    out.p_ci_low = std::max(0.0, out.p_hat - z * se_p);
    out.p_ci_high = std::min(1.0, out.p_hat + z * se_p);

    out.time_mean = sum_time / static_cast<double>(N);
    out.energy_mean = sum_energy / static_cast<double>(N);

    if (N > 1) {
        const double var_time =
            (sum_time2 - static_cast<double>(N) * out.time_mean * out.time_mean)
            / static_cast<double>(N - 1);
        const double var_energy =
            (sum_energy2 - static_cast<double>(N) * out.energy_mean * out.energy_mean)
            / static_cast<double>(N - 1);
        out.time_std = std::sqrt(std::max(0.0, var_time));
        out.energy_std = std::sqrt(std::max(0.0, var_energy));
    }

    return out;
}

int main() {
    std::vector<EpisodeMetrics> log;
    // TODO: fill log from your simulator or robot runs
    EvalStats stats = compute_stats(log);
    std::cout << "Success rate: " << stats.p_hat
              << " CI95: [" << stats.p_ci_low
              << ", " << stats.p_ci_high << "]\n";
    return 0;
}
      

File: Advanced_Robotics_Course/Chapter20/Lesson5/Chapter20_Lesson5.java
public class EpisodeMetrics {
    public int success;
    public double time;
    public double energy;
    public EpisodeMetrics(int success, double time, double energy) {
        this.success = success;
        this.time = time;
        this.energy = energy;
    }
}

public class EvalStats {
    public double pHat, pCiLow, pCiHigh;
    public double timeMean, timeStd;
    public double energyMean, energyStd;
}

public class Evaluator {

    public static EvalStats computeStats(java.util.List<EpisodeMetrics> data) {
        int N = data.size();
        double sumSuccess = 0.0;
        double sumTime = 0.0, sumTime2 = 0.0;
        double sumEnergy = 0.0, sumEnergy2 = 0.0;

        for (EpisodeMetrics e : data) {
            sumSuccess += (double)e.success;
            sumTime += e.time;
            sumTime2 += e.time * e.time;
            sumEnergy += e.energy;
            sumEnergy2 += e.energy * e.energy;
        }

        EvalStats out = new EvalStats();
        out.pHat = sumSuccess / (double)N;

        double z = 1.96;
        double seP = Math.sqrt(out.pHat * (1.0 - out.pHat) / (double)N);
        out.pCiLow = Math.max(0.0, out.pHat - z * seP);
        out.pCiHigh = Math.min(1.0, out.pHat + z * seP);

        out.timeMean = sumTime / (double)N;
        out.energyMean = sumEnergy / (double)N;

        if (N > 1) {
            double varTime =
                (sumTime2 - (double)N * out.timeMean * out.timeMean) / (double)(N - 1);
            double varEnergy =
                (sumEnergy2 - (double)N * out.energyMean * out.energyMean) / (double)(N - 1);
            out.timeStd = Math.sqrt(Math.max(0.0, varTime));
            out.energyStd = Math.sqrt(Math.max(0.0, varEnergy));
        }

        return out;
    }

    public static void main(String[] args) {
        java.util.List<EpisodeMetrics> log = new java.util.ArrayList<>();
        // TODO: populate log from your experiment runner
        EvalStats stats = computeStats(log);
        System.out.println("Success rate: " + stats.pHat +
                           " CI95: [" + stats.pCiLow + ", " + stats.pCiHigh + "]");
    }
}
      
